### PR TITLE
Standardise CI & Makefile across Miden repos

### PR DESCRIPTION
**Pull request by phklive**
In this PR I propose a standardisation of the CI and Makefile across Miden repositories. 

With the following changes: 
- Cargo make to Make
- CI using Cargo make to Make

Once this first one is approved I will port it to the following repos:
- Crypto
- Client
- Compiler

### CODE REVIEW

**Comment by bobbinth** [comment]
Let's hold off on this until we figure out how to deal with `async` features.

**Comment by bobbinth** [comment]
I think we can pick this up again now. This would include:

- Update the Makefile to be similar to the one in `miden-base` (but accounting for some unique tasks here).
  - As a side effect, we'll also need to switch to `nextest` here.
  - This will also require updating the docs (e.g., [here](https://0xpolygonmiden.github.io/miden-vm/intro/usage.html#cli-interface)) and potentially in other places.
- Update the CI to be similar to the one in `miden-base`.
  - This may also require updating badges in the main readme.
- Update the CONTRIBUTING file (probably just a few minor updates).

**Comment by phklive** [comment]
> I think we can pick this up again now. This would include:
> 
> * Update the Makefile to be similar to the one in `miden-base` (but accounting for some unique tasks here).
>   
>   * As a side effect, we'll also need to switch to `nextest` here.
>   * This will also require updating the docs (e.g., [here](https://0xpolygonmiden.github.io/miden-vm/intro/usage.html#cli-interface)) and potentially in other places.
> * Update the CI to be similar to the one in `miden-base`.
>   
>   * This may also require updating badges in the main readme.
> * Update the CONTRIBUTING file (probably just a few minor updates).

Absolutely, working on it.

**Comment by phklive** [comment]
@bobbinth should we still have a CI test on windows2022?

**Comment by phklive** [comment]
In the VM workspace rust version is 1.75 everywhere except in the `assembly` where it's 1.76. 

Should we bump everything to 1.78?

**Comment by phklive** [comment]
@bobbinth, I tried compiling with metal but I am receiving a large number of errors, I followed the instructions in the documentation I am not sure if I am doing something wrong or if the metal compilation is broken.

**Comment by bobbinth** [comment]
> In the VM workspace rust version is 1.75 everywhere except in the `assembly` where it's 1.76.
> 
> I bumped everything to rust 1.78 and standardised all the `cargo.toml`s. Let me know if I should revert.

1.78 sounds good - thank you! I think we'll probably skip 1.79 and will go straight to 1.80 once that's released in a couple of months.

> @bobbinth, I tried compiling with metal but I am receiving a large number of errors, I followed the instructions in the documentation I am not sure if I am doing something wrong or if the metal compilation is broken.

@TheMenko will make a PR to fix it soon.

**Comment by bobbinth** [comment]
> @bobbinth I removed the tests on Windows2022, let me know if we should add them back.

Unless they complicate things too much, I'd prefer to keep them (would be good to make sure the VM works fine on Windows machines too). iirc, these were run only when merging into `main`.

**Comment by phklive** [comment]
> > In the VM workspace rust version is 1.75 everywhere except in the `assembly` where it's 1.76.
> > I bumped everything to rust 1.78 and standardised all the `cargo.toml`s. Let me know if I should revert.
> 
> 1.78 sounds good - thank you! I think we'll probably skip 1.79 and will go straight to 1.80 once that's released in a couple of months.
> 
> > @bobbinth, I tried compiling with metal but I am receiving a large number of errors, I followed the instructions in the documentation I am not sure if I am doing something wrong or if the metal compilation is broken.
> 
> @TheMenko will make a PR to fix it soon.

Should we wait for his PR to merge this? 

This breaks all the `--all-features` commands. Which was requested by @plafer.

File: docs/src/intro/usage.md [hunk 1]
```diff
@@ -1,124 +1,156 @@
...
 ```
+
 For a faster build, you can compile with less optimizations, replacing `--profile optimized` by `--release`. Example:
+
 ```
-cargo build --release --features concurrent,executable
+make build-release
 ```
+
 In this case, the `miden` executable will be placed in the `./target/release` directory.
```
Comment (bobbinth) [inline docs/src/intro/usage.md:41]: I think we can probably get rid of this option.

File: miden/README.md
Comment (bobbinth) [inline miden/README.md:1]: We also need to update the [CLI Interface](https://github.com/0xPolygonMiden/miden-vm/tree/next/miden#cli-interface) section in this file.

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you! I left a few comments inline.

Also, do we not need a config file for `nextest`? (i.e., something similar to what we have in [miden-base](https://github.com/0xPolygonMiden/miden-base/blob/next/.config/nextest.toml)).

Lastly, let's update the changelog.

**Comment by bobbinth** [comment]
> > > In the VM workspace rust version is 1.75 everywhere except in the `assembly` where it's 1.76.
> > > I bumped everything to rust 1.78 and standardised all the `cargo.toml`s. Let me know if I should revert.
> > 
> > 
> > 1.78 sounds good - thank you! I think we'll probably skip 1.79 and will go straight to 1.80 once that's released in a couple of months.
> > > @bobbinth, I tried compiling with metal but I am receiving a large number of errors, I followed the instructions in the documentation I am not sure if I am doing something wrong or if the metal compilation is broken.
> > 
> > 
> > @TheMenko will make a PR to fix it soon.
> 
> Should we wait for his PR to merge this?
> 
> This breaks all the `--all-features` commands. Which was requested by @plafer.

Yes, let's wait for #1357 to be merged.

Regarding `--all-features` - is this because of the `metal` feature? How did it work before (i.e., we had this feature before this PR)?

**Comment by phklive** [comment]
> @phklive - now that #1357 has been merged, let's pick this up again.

Will do.

File: Makefile [hunk 2]
```diff
@@ -1,27 +1,86 @@
...
-	RUSTFLAGS="-C target-feature=+avx2" cargo build $(PROFILE_OPTIMIZED) $(FEATURES_CONCURRENT_EXEC)
+.PHONY: build-sve
+build-sve: ## Builds VM for sve
+	RUSTFLAGS="-C target-feature=+sve" cargo build --profile optimized $(FEATURES_CONCURRENT_EXEC)
 
-exec-sve:
-	RUSTFLAGS="-C target-feature=+sve" cargo build $(PROFILE_OPTIMIZED) $(FEATURES_CONCURRENT_EXEC)
+.PHONY: build-info
+build-info: ## Builds VM with log tree
+	cargo build --profile optimized $(FEATURES_LOG_TREE)
```
[the remark this answers is not in the document]
Comment (phklive) [reply Makefile:80]: I don't fully understand why they are called `exec` and not build considering that the underlying action that is performed is building.

[the remark this answers is not in the document]
Comment (phklive) [reply Makefile:80]: How do we differentiate the `build` and `exec` commands?

File: Makefile [hunk 3]
```diff
@@ -1,27 +1,96 @@
...
+
+# --- docs ----------------------------------------------------------------------------------------
+
+.PHONY: doc
+doc: ## Generates & checks documentation
+	$(WARNINGS) cargo doc --all-features --keep-going --release
+
+.PHONY: mdbook
+mdbook: ## Generates mdbook
+	mdbook exec docs/
```
Comment (bobbinth) [inline Makefile:54]: Is `mdbook exec` a valid command? I thought it was `mdbook build` (i.e., based on [this](https://rust-lang.github.io/mdBook/cli/index.html)).

Also, I'd write the comment as "Generated mdBook documentation".

File: Makefile [hunk 4]
```diff
@@ -1,27 +1,96 @@
...
+
+.PHONY: build
+build: ## Builds with default parameters
+	cargo build --release --features concurrent
+
+.PHONY: build-no-std
+build-no-std: ## Builds without the standard library
+	cargo build --no-default-features --target wasm32-unknown-unknown --workspace
+
+# --- executing ------------------------------------------------------------------------------------
```
Comment (bobbinth) [inline Makefile:66]: nit: I would name this section "building an executable" or maybe just "executable".

File: Makefile [hunk 5]
```diff
@@ -1,27 +1,96 @@
...
+
+.PHONY: exec
+exec: ## Builds with optimized profile and features
```
Comment (bobbinth) [inline Makefile:69]: nit: I would change the comment to "Builds an executable with optimized profile and features". I would also make similar changes to other such commands.

File: Makefile [hunk 6]
```diff
@@ -1,27 +1,96 @@
...
+	cargo build --profile optimized $(FEATURES_CONCURRENT_EXEC)
+
+.PHONY: exec-single
+exec-single: ## Builds in single-threaded mode
+	cargo build --profile optimized --features executable
 
-exec:
-	cargo build $(PROFILE_OPTIMIZED) $(FEATURES_CONCURRENT_EXEC)
+.PHONY: exec-metal
+exec-metal: ## Builds for metal
```
Comment (bobbinth) [inline Makefile:77]: I would change the comment here to "Builds an executable with Metal acceleration enabled"

**Comment by phklive** [comment]
> @bobbinth or @bitwalker to make sure that the Miden VM passes `cargo doc` I will need your help concerning this piece of documentation:
> 
> https://github.com/0xPolygonMiden/miden-vm/blob/bf885b7bf306d283c23dba44efe0dbd2db11efe6/assembly/src/assembler/mod.rs#L61-L85
> 
> Multiple comments have been made referencing non existing methods:
> 
> * `Assembler::with_kernel_from_source`
> * `Assembler::compile`
> * `Assembler::compile_ast`
> * `Assembler::compile_with_opts`
> 
> And here:
> 
> https://github.com/0xPolygonMiden/miden-vm/blob/bf885b7bf306d283c23dba44efe0dbd2db11efe6/assembly/src/assembler/context.rs#L21
> 
> * `crate::Assembler::compile_in_context`
> 
> I am sure we can easily fix these issues considering they should be pointing to existing code, I just need your support to understand which pieces of code you meant when writing the docs.

@bobbinth, @bitwalker could you give me a hand on this one. 

It's mainly the only thing we need before being able to PR.

**Comment by bobbinth** [comment]
> @bobbinth, @bitwalker could you give me a hand on this one.

I pushed a small commit which should fix the doc build.

**Comment by phklive** [comment]
I have been facing errors while trying to run the tests on `Windows`:  https://github.com/0xPolygonMiden/miden-vm/actions/runs/9963373945/job/27529273724?pr=1342#step:5:272

To make sure that we can merge this PR without waiting more I have opened an issue to solve these problems: https://github.com/0xPolygonMiden/miden-vm/issues/1402

File: Makefile [hunk 7]
```diff
@@ -1,27 +1,96 @@
...
+
+
+.PHONY: lint
+lint: format fix clippy ## Runs all linting tasks at once (Clippy, fixing, formatting)
+
+# --- testing -------------------------------------------------------------------------------------
+
+.PHONY: test
+test: ## Runs all tests
+	$(DEBUG_ASSERTIONS) cargo nextest run --cargo-profile test-release --features internals
```
Comment (bobbinth) [inline Makefile:44]: Could we move the "testing" section after "docs". I think that's the order we have in `miden-base`.

Also, we are missing `check` target (which runs `cargo check`). I'd add it here to keep things consistent with `miden-base`.

File: .github/workflows/changelog.yml
Comment (bobbinth) [inline .github/workflows/changelog.yml:1]: We probably shouldn't delete this file?

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        toolchain: [stable, nightly]
+    steps:
+      - uses: actions/checkout@main
+      - name: Build for avx2
+        run: |
+          rustup update --no-self-update ${{ matrix.toolchain }}
+          make exec-avx2
```
Comment (bobbinth) [inline .github/workflows/build.yml:51]: I'm actually not sure if these are needed as the CI machine is unlikely to have the required hardware to build these properly (so, they will most likely just default to a standard build).

File: README.md [hunk 9]
```diff
@@ -1,163 +1,183 @@
...
 
-<a href="https://github.com/0xPolygonMiden/miden-vm/blob/main/LICENSE"><img src="https://img.shields.io/badge/license-MIT-blue.svg"></a>
-<img src="https://github.com/0xPolygonMiden/miden-vm/workflows/CI/badge.svg?branch=main">
-<img src="https://img.shields.io/badge/rustc-1.75+-lightgray.svg">
-<a href="https://crates.io/crates/miden-vm"><img src="https://img.shields.io/crates/v/miden-vm"></a>
+[![LICENSE](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/0xPolygonMiden/miden-vm/blob/main/LICENSE)
+[![test](https://github.com/0xPolygonMiden/miden-vm/actions/workflows/test.yml/badge.svg)](https://github.com/0xPolygonMiden/miden-vm/actions/workflows/test.yml)
+[![build](https://github.com/0xPolygonMiden/miden-vm/actions/workflows/build.yml/badge.svg)](https://github.com/0xPolygonMiden/miden-vm/actions/workflows/build.yml)
+[![RUST_VERSION](https://img.shields.io/badge/rustc-1.78+-lightgray.svg)](https://www.rust-lang.org/tools/install)
+[![GitHub Release](https://img.shields.io/github/release/0xPolygonMiden/miden-vm)](https://github.com/0xPolygonMiden/miden-vm/releases/)
```
Comment (bobbinth) [inline README.md:7]: I think this is redundant with the next line. So, I'd remove it.

File: miden/README.md [hunk 10]
```diff
@@ -32,7 +32,7 @@ Having a small number elements to describe public inputs and outputs of a progra
 
 ## Usage
 
-Miden crate exposes several functions which can be used to execute programs, generate proofs of their correct execution, and verify the generated proofs. How to do this is explained below, but you can also take a look at working examples [here](examples) and find instructions for running them via CLI [here](#fibonacci-example).
+Miden crate exposes several functions which can be used to execute programs, generate proofs of their correct execution, and verify the generated proofs. How to do this is explained below, but you can also take a look at working examples [here](src/examples) and find instructions for running them via CLI [here](#fibonacci-example).
```
Comment (bobbinth) [inline miden/README.md:35]: I think the reference to `examples` here was correct (we want the examples directory in the root of the crate rather in in `src`.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! I left a few comments inline - but all of them are pretty small.

File: .github/workflows/changelog.yml
Comment (bobbinth) [reply .github/workflows/changelog.yml:1]: Outdated - please ignore.

File: README.md (hunk 9 above)
Comment (phklive) [reply README.md:7]: I left both because the github release here is not the same as the crates io release.

Comment (bobbinth) [reply README.md:7]: I think that's OK - having two versions may be more confusing that helpful.

**Pull request merged**